homeassistant/components/tasmota/switch.py

```diff
@@ -0,0 +1,132 @@
+"""Support for Tasmota switches."""
+import logging
+
+from homeassistant.components import switch
+from homeassistant.components.switch import SwitchEntity
+from homeassistant.core import callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+
+from . import DOMAIN as TASMOTA_DOMAIN
+from .discovery import TASMOTA_DISCOVERY_ENTITY_NEW, clear_discovery_hash
+from .mixins import TasmotaAvailability, TasmotaDiscoveryUpdate
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Tasmota switch dynamically through discovery."""
+
+    async def async_discover(entity, discovery_hash):
+        """Discover and add a Tasmota switch."""
+        try:
+            await _async_setup_entity(
+                entity, discovery_hash, async_add_entities, config_entry
+            )
+        except Exception:
+            clear_discovery_hash(hass, discovery_hash)
+            raise
+
+    async_dispatcher_connect(
+        hass,
+        TASMOTA_DISCOVERY_ENTITY_NEW.format(switch.DOMAIN, TASMOTA_DOMAIN),
+        async_discover,
+    )
+
+
+async def _async_setup_entity(entity, discovery_hash, async_add_entities, config_entry):
```

Comment:
This can be a callback instead of a coroutine function since we don't await inside.

Reply:
I removed the function altogether, it was a bit pointless